internal/migrations/003-primitive-insertion.sql

```diff
@@ -80,6 +81,13 @@ CREATE OR REPLACE ACTION insert_records(
         }
     }
 
+    -- Get stream reference for all streams
+    $stream_refs := []::UUID[];
+    for $i in 1..array_length($data_provider) {
+      $id := get_stream_id($data_provider[$i], $stream_id[$i]);
```

Comment:
Although this works, usually there is a significant performance hit with this kind of loop...

What happens is that the for loop here occurs at the interpreter (golang) while the query inside happens at PostgreSQL, and this interface roundtripping has some overhead that happens in loops

So a good measure is to have a batch version of `get_stream_id` to abstract that which could happen at PostgreSQL all at once

Another thing that this enables is using `DISTINCT ON`, which can remove the need for getting the same refs lookup more than once

For example, we usually insert 10K~25K records at once, sometimes for a few distinct streams, but many times for the same stream

but as this is performance only, if we prefer to approve and try to address in another PR would be good too